### PR TITLE
Add explanation for our docstring in API reference

### PR DESCRIPTION
**Pull request by himkt**
This PR is related to https://github.com/optuna/optuna/issues/2978.

Our documentation enumerates exceptions that users might encounter for each API. However, it is impossible or really difficult to cover all the exceptions because an exception could be raised in third-party libraries that we use in Optuna.

Given the situation, it would need to note that undocumented exceptions are not specification somewhere (refer to the issue for more information). I propose to note it in the index page of API reference with explaining our docstring style (and also note that continuously improvement is contribution-welcome).

### CODE REVIEW

**Comment by himkt** [comment]
The CI failures will be fixed in https://github.com/optuna/optuna/pull/3308.

**Comment by himkt** [comment]
Could you please give me a review @hvy @HideakiImamura?

File: docs/source/reference/index.rst [hunk 1]
```diff
@@ -1,6 +1,55 @@
 API Reference
 =============
 
+
+How to read Optuna API documentation?
+-------------------------------------
```
Comment (hvy) [inline docs/source/reference/index.rst:6]:
```suggestion
How to read the API documentation?
----------------------------------
```

File: docs/source/reference/index.rst [hunk 2]
```diff
@@ -1,6 +1,55 @@
...
+
+We provide the API reference that is automatically generated by
+docstrings written in each modules. We basically follow `Google Style Python Docstrings <https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html>`_.
+For more information, please read our `Coding Style Conventions <https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#docstrings>`_ as well.
+A typical example of our docstring looks like following:
+
+.. code-block:: python
+
+    def hoge(self, number: int, ...) -> int:
```
Comment (hvy) [inline docs/source/reference/index.rst:15]: A minor suggestion. `foo` or perhaps something slightly more verbose might be better. `hoge` is quite domestic to Japan.
```suggestion
    def some_function(self, number: int, ...) -> int:
```

File: docs/source/reference/index.rst [hunk 3]
```diff
@@ -1,6 +1,55 @@
...
+        """Description of a method
```
Comment (hvy) [inline docs/source/reference/index.rst:16]:
```suggestion
        """Description of a method.
```

File: docs/source/reference/index.rst [hunk 4]
```diff
@@ -1,6 +1,55 @@
...
+
+        More detailed explanation for the API...
+        ...
+
+        Example:
+
+            .. testcode::
+
+                your_awesome_number = 1
+                hoge(your_awesome_number)
```
Comment (hvy) [inline docs/source/reference/index.rst:26]:
```suggestion
                some_function(your_awesome_number)
```

File: docs/source/reference/index.rst [hunk 5]
```diff
@@ -1,6 +1,55 @@
...
+
+                ...
+
+        Args:
+            number:
+                Some description for the argument here
```
Comment (hvy) [inline docs/source/reference/index.rst:32]:
```suggestion
                Some description for the argument here.
```

File: docs/source/reference/index.rst [hunk 6]
```diff
@@ -1,6 +1,55 @@
...
+            ...
+
+        Raises:
+            TypeError:
+                Describe here when an exception is raised.
```
Comment (hvy) [inline docs/source/reference/index.rst:37]:
```suggestion
                Describe here when an exception is raised.
```

Comment (hvy) [inline docs/source/reference/index.rst:37]: This explanatory text seems to be targeting developers. How about targeting users as it's part of the library documentation? E.g. "Description of common error."

File: docs/source/reference/index.rst [hunk 7]
```diff
@@ -1,6 +1,55 @@
 API Reference
 =============
 
+
+How to read Optuna API documentation?
+-------------------------------------
+
+We provide the API reference that is automatically generated by
+docstrings written in each modules. We basically follow `Google Style Python Docstrings <https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html>`_.
+For more information, please read our `Coding Style Conventions <https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#docstrings>`_ as well.
```
Comment (hvy) [inline docs/source/reference/index.rst:10]: Some grammatical changes and simplifications. What do you think?
```suggestion
The API reference style is based on `Google Style Python Docstrings <https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html>`_.
For more information, please refer to the `Coding Style Conventions <https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#docstrings>`_.
```

File: docs/source/reference/index.rst [hunk 8]
```diff
@@ -1,6 +1,55 @@
...
+
+        Raises:
+            TypeError:
+                Describe here when an exception is raised.
+            ...
+
+
+Each documentation shows information about `Parameters` (Args), `Raises` and `Return type`.
+Types of parameters and return value are based on type hints, which are annotated to each methods in our codebase.
+It also contains `Methods` and `Attributes` in the class documentation and some API shows a code example for you.
```
Comment (hvy) [inline docs/source/reference/index.rst:43]: Ditto.

**Review by hvy — COMMENTED** [review]
Thanks a lot for the PR. I took a quick look and left some early comments.

I'd just like to clarify if we really want to base the explanation on the source code. The API references are generated results of those (as you've included a description in the PR), and not the source code itself. In other words, it looks a bit like it's targeting developers, and not users. But I understand that the explanation about the coverage of the exceptions in the API references is very informative. Maybe this is a good first approach!

File: docs/source/reference/index.rst (hunk 7 above)
Comment (hvy) [reply docs/source/reference/index.rst:10]: This also seems to target developers and not users.

File: docs/source/reference/index.rst (hunk 2 above)
Comment (himkt) [reply docs/source/reference/index.rst]: I didn't know that..., thank you for letting me know! :memo: And adding `_function` also improve the readability. 🙏

File: docs/source/reference/index.rst (hunk 3 above)
Comment (himkt) [reply docs/source/reference/index.rst]: :bow:

**Review by himkt — COMMENTED** [review]
Thank you for the comments. As you pointed out, I notice that I wrote the docs without taking care of targets. I may not change that it uses docstring (API reference) but I'll update a explanation to be informative for library users. 🚧

**Comment by himkt** [comment]
@hvy I tried to rewrite the description to be more targeting library users. I'd like to hear your impression. :pray:

**Comment by himkt** [comment]
@hvy suggested me that this information should be documented in wiki. I close this PR and update our wiki developer guideline. After doing that, I'll open another PR to update API reference if it needs.

**Pull request closed without merging**